Sources/NIOOpenSSL/OpenSSLHandler.swift

```diff
@@ -248,27 +264,46 @@ public class OpenSSLHandler : ChannelInboundHandler, ChannelOutboundHandler {
     /// Loops over the `SSL` object, decoding encrypted application data until there is
     /// no more available.
     private func doDecodeData(ctx: ChannelHandlerContext) {
+        guard var receiveBuffer = self.plaintextReadBuffer else {
+            preconditionFailure("didDecodeData called without handlerAdded firing.")
+        }
+
+        // We nil the read buffer here. This is done on purpose: we do it to ensure
+        // that we don't have two references to the buffer, otherwise readDataFromNetwork
+        // will trigger a CoW every time. We need to put this back on every exit from this
+        // function, or before any call-out, to avoid re-entrancy issues.
+        self.plaintextReadBuffer = nil
+
         readLoop: while true {
-            let result = connection.readDataFromNetwork(allocator: ctx.channel.allocator)
+            let result = connection.readDataFromNetwork(outputBuffer: &receiveBuffer)
             
             switch result {
-            case .complete(let buf):
-                // TODO(cory): Should we coalesce these instead of dispatching multiple times? I think so!
-                // It'll also let us avoid this weird boolean flag, which is always good.
-                didDeliverData = true
-                ctx.fireChannelRead(self.wrapInboundOut(buf))
+            case .complete:
+                // Good read. We need to check how many bytes are still writable and, if it's
+                // too few, move them along. As a general heuristic we want room to write 1kB
+                // of data, though this number is utterly arbitrary.
+                if receiveBuffer.writableBytes < 1024 {
+                    receiveBuffer.changeCapacity(to: receiveBuffer.capacity * 2)
```

Comment:
- this doesn't guarantee it's `>= 1024` still, right? Maybe `changeCapacity(to: 1025)`?
- also shouldn't we use `reserveCapacity` and just require NIO >= 1.9.0? People still have the old versions available if they want to stay on the old stuff

Reply:
I think it does. The buffer is originally allocated at 16kB, so doubling the capacity will always yield a buffer at least twice as large.

Reply:
I can assert to that effect if you want.

Reply:
after chatting to you I suggest `receiveBuffer.capacity + 1024`